### PR TITLE
migration lint for `expr2024` for the edition 2024

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
This is adding a migration lint for the current (in the 2021 edition and previous)
to move expr to expr_2021 from expr

Issue https://github.com/rust-lang/rust/issues/123742

I created also a repository to test out the migration https://github.com/vincenzopalazzo/expr2024-cargo-fix-migration

Co-Developed-by: @eholk

### CODE REVIEW

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (traviscross) [inline compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:22]:
```suggestion
    To keep the existing behavior, use the `expr_2021` fragment specifier.",
```

File: compiler/rustc_lint/src/lints.rs
Comment (traviscross) [inline compiler/rustc_lint/src/lints.rs:328]: Presumably we want to name this...

```suggestion
pub struct MacroExprFragment2024 {
```

...with corresponding updates elsewhere.

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (vincenzopalazzo) [inline compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:19]: I am not good at writing documentation, @traviscross  or @eholk have some suggestions on what to write here?

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (eholk) [inline compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:51]: We probably shouldn't check this in with a FIXME here. Also, it's `expr_2021` (with the `_`) and not `expr2021` (without the `_`), so we need to be consistent.

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (eholk) [inline compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:70]: I wonder if it's better to link to RFC 3531?

File: tests/ui/macros/expr_2021_cargo_fix_edition.rs
Comment (eholk) [inline tests/ui/macros/expr_2021_cargo_fix_edition.rs:6]: It might be good to add another clause here that has an `expr` deeper in the pattern to make sure we properly recurse and all that.

**Review by eholk — COMMENTED** [review]
This seems to be in a pretty decent shape, but I left a few comments.

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (traviscross) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:70]: Linking to the page in the edition guide would be the best.

File: tests/ui/macros/expr_2021_cargo_fix_edition.rs
Comment (vincenzopalazzo) [reply tests/ui/macros/expr_2021_cargo_fix_edition.rs:6]: Ah make really sense!

Comment (vincenzopalazzo) [reply tests/ui/macros/expr_2021_cargo_fix_edition.rs:6]: Done in https://github.com/rust-lang/rust/pull/125627/commits/facc35c75c697a5b4dca3a2a78fdd9817201f19e

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: We do not have a real edition guide for this @traviscross, or are you saying just to point the user to the edition guide?

At the moment I added the RFC that @eholk, but we can add also the edition guide if you like.

Trying to address in the commit d665cd868a0d6ecfec0d07acbe8de10f3b879772****

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:51]: I suggest removing the 'fixme' for now since the feature is not yet finished. We can rework it once we have the full picture.

Let me know what you think.

Comment (eholk) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:51]: Yeah, I think it's fine to just remove the FIXME.

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (eholk) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: I agree that the edition guide would be best but first we have to write that section of the edition guide :)

Why don't we link the RFC for now and update it with a link to the edition guide once we have it?

Comment (traviscross) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: On the edition guide, if you want to put up PR just for a stub page (that itself for now just links elsewhere), so that can be linked here, we can merge that.  E.g., [here's an existing stub](https://doc.rust-lang.org/stable/edition-guide/rust-2024/rustfmt-overflow-delimited-expr.html).  (I'd put it up myself but then I couldn't merge it.)

Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: Thanks, I think makes sense, we can do it

Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: OK added the migration guide, let me know if you want change the wording before I will squash the commtis

**Comment by vincenzopalazzo** [comment]
This should be the final iteration @eholk (modulo some wording change inside the docs)

Giving another run to the CI, hoping that now it is fine with my changes

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (traviscross) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: See: https://github.com/rust-lang/rust/pull/125627#discussion_r1630089274

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:67]: Ok, I am reverting the change and linking at the RFC, because we still need to write the `_` PR and then we go to the migration guide.

So I think there is no reason to rework this section if this need to be change anyway

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: Linking the RFC for now, then then we arrive at the doc step we rework this section, putting a note on my todo list

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (eholk) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: I put up a PR for a stub here: https://github.com/rust-lang/edition-guide/pull/302

So once that merges, we can link to that page from here.

Comment (traviscross) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: It's merged.

Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: Mh looks like I am not able to find the link? https://doc.rust-lang.org/stable/edition-guide/introduction.html

There is some other kind of edition version? like nightly?

Comment (traviscross) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: It's just waiting on a submodule update like [this one](https://github.com/rust-lang/rust/pull/125933) to happen.  It's OK to use the URL in this PR before it goes live.  It'll be:

https://doc.rust-lang.org/stable/edition-guide/rust-2024/macro-fragment-specifiers.html

File: tests/ui/macros/expr_2021_cargo_fix_edition.stderr
Comment (traviscross) [inline tests/ui/macros/expr_2021_cargo_fix_edition.stderr:30]: (Line picked semi-randomly...)

For this change, we should probably have one or more `run-rustfix` tests.  See:

https://rustc-dev-guide.rust-lang.org/tests/ui.html?highlight=rustfix#rustfix-tests

File: compiler/rustc_lint/src/lints.rs
Comment (traviscross) [inline compiler/rustc_lint/src/lints.rs:331]: For this change, we should probably have one or more `run-rustfix` tests.  See:

https://rustc-dev-guide.rust-lang.org/tests/ui.html?highlight=rustfix#rustfix-tests

**Comment by compiler-errors** [comment]
This incorrectly suggests changing the literal identifier `expr` in macros:

```rust
macro_rules! test {
    (expr) => {}
}

test!(expr);
```

which suggests:

```
help: to keep the existing behavior, use the `expr_2021` fragment specifier.
  |
LL |     (expr_2021) => {}
  |      ~~~~~~~~~
```

I believe that should be fixed.

I think a quick fix would be to just look for a preceding `:` token rather than the lack of a `$` token, but I wonder if there's a better way of getting the list of macro arguments from a `macro_rules` definition.

@bors r-

File: compiler/rustc_lint/messages.ftl
Comment (compiler-errors) [inline compiler/rustc_lint/messages.ftl:445]:
```suggestion
lint_macro_expr_fragment_specifier_2024_migration =
    the `expr` fragment specifier will accept more expressions in the 2024 edition
    .suggestion = to keep the existing behavior, use the `expr_2021` fragment specifier
```

Diagnostic fragments don't have `.` at the end of them

**Comment by compiler-errors** [comment]
Actually, a more resilient solution would be to maintain 3 tokens of lookback, so we can look for `$`, followed by an identifier, followed by `:`, and then followed by `expr`.

Then we don't have to worry about false positives like:

```
macro_rules! e {
    ($e:tt : expr) => {}
}
```

**Comment by vincenzopalazzo** [comment]
>I think a quick fix would be to just look for a preceding : token rather than the lack of a $ token, but I wonder if there's a better way of getting the list of macro arguments from a macro_rules definition.

Uh! Thanks for the catch, @compiler-errors. I overlooked the parsing code. Patching it now with your suggestion by checking the previous token if it matches :.

However, I think it is worth mentioning that @eholk and I spent some time looking into whether it was possible to use the Rustc parser macro inside Cargo fix. It seemed like it should be possible, but at the time, if I recall correctly, it was not trivial, and we ended up doing this manual parsing.

Maybe we should look inside this with the @rust-lang/wg-macros

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (traviscross) [inline compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:60]: This language seems to have been copy and pasted from something else and not updated for what this lint does.  E.g., it talks about switching code to use raw identifiers.

**Review by traviscross — CHANGES_REQUESTED** [review]

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: oh good catch, thanks. I think why we need this lint need to be standard across all the cargo fix lint for this reason I try to not rewrite the same thing in other words.

I am removing the link and the pointers to the raw identifier feature. Let me know if we need to improve it in any other way

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (traviscross) [inline compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs:59]: This needs further changes, e.g. to explain what this lint does and what people should know about it (e.g. that they probably want to switch back to `expr` from `expr_2021`), and to remove text from the explanation that is not correct for what this lint does.

**Review by traviscross — CHANGES_REQUESTED** [review]

**Comment by eholk** [comment]
> This incorrectly suggests changing the literal identifier `expr` in macros:
> 
> ```rust
> macro_rules! test {
>     (expr) => {}
> }
> 
> test!(expr);
> ```
> 
> which suggests:
> 
> ```
> help: to keep the existing behavior, use the `expr_2021` fragment specifier.
>   |
> LL |     (expr_2021) => {}
>   |      ~~~~~~~~~
> ```

It'd be good to add a test case to make sure we don't do this.

File: compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs
Comment (vincenzopalazzo) [reply compiler/rustc_lint/src/macro_expr_fragment_specifier_2024_migration.rs]: @traviscross we update the docs with @eholk so should be good for another review

**Comment by vincenzopalazzo** [comment]
@compiler-errors When you have time I would like to have another review on the parsing code to see if the code can have your bless 😸

**Comment by compiler-errors** [comment]
This looks fine to me for now. Please rebase to squash that `fixup!` commit, since bors doesn't autosquash.

**Comment by vincenzopalazzo** [comment]
>This looks fine to me for now. Please rebase to squash that fixup! commit, since bors doesn't autosquash.

Concurrency problem, I did while you was writing the comment sorry 😸

**Review by compiler-errors — APPROVED** [review]

**Review by traviscross — APPROVED** [review]

**Pull request merged**
